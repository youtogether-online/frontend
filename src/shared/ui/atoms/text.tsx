import { styled } from '@stitches/react'
import { ReactNode } from 'react'

type TypographyVariant =
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6'
  | 'subtitle1'
  | 'subtitle2'
  | 'body1'
  | 'body2'
  | 'caption'
  | 'keyboard'
  | 'button'
  | 'link'

type TextHTMLElements =
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6'
  | 'span'
  | 'p'
  | 'kbd'

interface VariantsMapping {
  [name: string]: TextHTMLElements
}

const variantsMapping: VariantsMapping = {
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h5: 'h5',
  h6: 'h6',
  subtitle1: 'h6',
  subtitle2: 'h6',
  body1: 'p',
  body2: 'p',
  button: 'span',
  link: 'span',
  caption: 'span',
  keyboard: 'kbd',
}

interface TypographyProps {
  variant: TypographyVariant
  children: ReactNode
  strong?: boolean
  secondary?: boolean
  centered?: boolean
}

export const Text = ({
  variant,
  children,
  strong = false,
  secondary = false,
  centered = false,
  ...textComponentProps
}: TypographyProps) => {
  const TextHTMlElement: TextHTMLElements = variantsMapping[variant]

  return (
    <TextStyled
      as={TextHTMlElement}
      variant={variant}
      strong={strong}
      secondary={secondary}
      centered={centered}
      {...textComponentProps}
    >
      {children}
    </TextStyled>
  )
}

const TextStyled = styled('p', {
  variants: {
    variant: {
      h1: {
        fontSize: '$h1',
        fontWeight: '$semiBold',
      },
      h2: {
        fontSize: '$h2',
        fontWeight: '$semiBold',
      },
      h3: {
        fontSize: '$h3',
        fontWeight: '$semiBold',
      },
      h4: {
        fontSize: '$h4',
        fontWeight: '$semiBold',
      },
      h5: {
        fontSize: '$h5',
        fontWeight: '$semiBold',
      },
      h6: {
        fontSize: '$h6',
        fontWeight: '$semiBold',
      },
      body1: {
        fontSize: '$body1',
        fontWeight: '$normal',
      },
      body2: {
        fontSize: '$body2',
        fontWeight: '$normal',
      },
      button: {
        fontSize: '$button',
        fontWeight: '$normal',
      },
      link: {
        fontSize: '$link',
        fontWeight: '$normal',
      },
      caption: {
        fontSize: '$caption',
        fontWeight: '$normal',
      },
      subtitle1: {
        fontSize: '$subtitle1',
        fontWeight: '$normal',
      },
      subtitle2: {
        fontSize: '$subtitle2',
        fontWeight: '$medium',
      },
      keyboard: {
        fontSize: '$keyboard',
        fontWeight: '$normal',
        border: '1px solid rgba(100,100,100, 0.2)',
        borderRadius: '3px',
        background: 'rgba(150, 150, 150, 0.06)',
        padding: '0px 4px',
        borderBottomWidth: '3px',
      },
    },
    strong: {
      true: {
        fontWeight: '$semiBold',
      },
    },
    centered: {
      true: {
        textAlign: 'center',
      },
    },
    secondary: {
      true: {
        color: '$textSecondary',
      },
    },
  },
})
