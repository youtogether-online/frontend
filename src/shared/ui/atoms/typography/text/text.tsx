import { cva, type VariantProps } from "cva";
import { type ReactNode } from "react";
import { tw } from "typewind";

import { baseStyles } from "../styles";

const text = cva("", {
  variants: {
    ...baseStyles,
    variant: {
      primary: [tw.text_fgDefault],
      secondary: [tw.text_fgMuted],
      danger: [tw.text_dangerFg],
    },
    component: {
      span: [tw.font_normal],
      p: [tw.font_normal],
      kbd: [
        tw.text_sm,
        tw.font_mono,
        tw.bg_canvasSubtle,
        tw.border_neutralMuted,
        tw.border,
        tw.border_solid,
        tw.rounded_md,
        tw.inline_block,
        tw.p_1,
      ],
      em: [],
      u: [],
      small: [],
      strong: [],
    },
  },
  defaultVariants: {
    component: "p",
    variant: "primary",
  },
});

type TextProps = VariantProps<typeof text> & {
  // Redefine `type` because cva doesn't provides way to set required props
  component: "span" | "p" | "kbd" | "em" | "u" | "small" | "strong";
  sx?: string;
  children: ReactNode;
};

export const Text = ({ component, children, variant }: TextProps) => {
  const Component = component;

  return <Component className={text({ component, variant })}>{children}</Component>;
};
