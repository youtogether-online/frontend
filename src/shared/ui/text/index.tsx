import { ReactNode } from "react";
import { styled } from "@stitches/react";

type TypographyVariant =
  "h1"
  | "h2"
  | "h3"
  | "h4"
  | "h5"
  | "h6"
  | "subtitle1"
  | "subtitle2"
  | "body1"
  | "body2"
  | "button"
  | "caption"
  | "keyboard"

type TextHTMLElements = "h1" | "h2" | "h3" | "h4" | "h5" | "h6" | "span" | "p" | "kbd"

type VariantsMapping = {
  [name: string]: TextHTMLElements
}

const variantsMapping: VariantsMapping = {
  h1: "h1",
  h2: "h2",
  h3: "h3",
  h4: "h4",
  h5: "h5",
  h6: "h6",
  subtitle1: "h6",
  subtitle2: "h6",
  body1: "p",
  body2: "p",
  button: "span",
  caption: "span",
  keyboard: "kbd"
};

type TypographyProps = {
  variant: TypographyVariant
  children: ReactNode
  strong?: boolean
  keyboard?: boolean
  secondary?: boolean
}
export const Text = ({
                       variant,
                       children,
                       strong = false,
                       keyboard = false,
                       secondary = false,
                       ...textComponentProps
                     }: TypographyProps) => {
  const TextHTMlElement: TextHTMLElements = variantsMapping[variant];

  return <TextStyled as={TextHTMlElement} variant={variant} strong={strong} secondary={secondary}
                     {...textComponentProps}>{children}</TextStyled>;
};

const TextStyled = styled("p", {
  variants: {
    variant: {
      h1: {
        fontSize: "$h1",
        fontWeight: "$semiBold"
      },
      h2: {
        fontSize: "$h2",
        fontWeight: "$semiBold"
      },
      h3: {
        fontSize: "$h3",
        fontWeight: "$semiBold"
      },
      h4: {
        fontSize: "$h4",
        fontWeight: "$semiBold"
      },
      h5: {
        fontSize: "$h5",
        fontWeight: "$semiBold"
      },
      h6: {
        fontSize: "$h6",
        fontWeight: "$semiBold"
      },
      body1: {
        fontSize: "$body1",
        fontWeight: "$normal"
      },
      body2: {
        fontSize: "$body2",
        fontWeight: "$normal"
      },
      button: {
        fontSize: "$button",
        fontWeight: "$normal",
        display: "block"
      },
      caption: {
        fontSize: "$caption",
        fontWeight: "$normal"
      },
      subtitle1: {
        fontSize: "$subtitle1",
        fontWeight: "$normal"
      },
      subtitle2: {
        fontSize: "$subtitle2",
        fontWeight: "$medium"
      },
      keyboard: {
        fontSize: "$keyboard",
        fontWeight: "$normal",
        border: "1px solid rgba(100,100,100, 0.2)",
        borderRadius: "3px",
        background: "rgba(150, 150, 150, 0.06)",
        padding: "0px 4px",
        borderBottomWidth: "3px"
      }
    },
    strong: {
      true: {
        fontWeight: "$semiBold"
      }
    },
    secondary: {
      true: {
        color: "$textSecondary"
      }
    }
  }
});

