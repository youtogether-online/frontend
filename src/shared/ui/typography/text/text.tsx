import { Slot } from "@radix-ui/react-slot";
import { type ReactNode } from "react";

import { type SxProp } from "../../types";

type TextProps = {
  asChild?: boolean;
  children: ReactNode;
} & SxProp;

export const Text = ({ asChild, children, sx }: TextProps) => {
  const Component = asChild ? Slot : "p";

  return <Component className={sx}>{children}</Component>;
};
