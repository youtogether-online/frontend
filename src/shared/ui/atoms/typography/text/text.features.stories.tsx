import type { Meta, StoryObj } from "@storybook/react";

import { Text } from "./text";

const meta: Meta<typeof Text> = {
  title: "Components/Typography/Text/Features",
  component: Text,
};

export default meta;
type Story = StoryObj<typeof Text>;

export const Variants: Story = {
  render: () => (
    <>
      <Text component="p" variant="primary">
        Primary
      </Text>
      <Text component="p" variant="secondary">
        Secondary
      </Text>
      <Text component="p" variant="danger">
        Danger
      </Text>
    </>
  ),
};

export const Components: Story = {
  render: () => (
    <div className="flex flex-col">
      <Text component="p">Parapgraph</Text>
      <Text component="kbd">Kbd</Text>
      <Text component="span">Span</Text>
      <Text component="em">Italized</Text>
      <Text component="small">Small</Text>
      <Text component="strong">Strong</Text>
      <Text component="u">Misspeled</Text>
    </div>
  ),
};
