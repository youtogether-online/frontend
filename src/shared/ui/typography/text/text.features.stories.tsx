import type { Meta, StoryObj } from "@storybook/react";

import { Text } from "./text";

const meta: Meta<typeof Text> = {
  title: "Components/Typography/Text/Features",
  component: Text,
};

export default meta;
type Story = StoryObj<typeof Text>;

export const Variants: Story = {
  render: () => (
    <>
      <Text as="p" variant="primary">
        Primary
      </Text>
      <Text as="p" variant="secondary">
        Secondary
      </Text>
      <Text as="p" variant="danger">
        Danger
      </Text>
    </>
  ),
};

export const Components: Story = {
  render: () => (
    <div className="flex flex-col">
      <Text as="p">Parapgraph</Text>
      <Text as="kbd">Kbd</Text>
      <Text as="span">Span</Text>
      <Text as="em">Italized</Text>
      <Text as="small">Small</Text>
      <Text as="strong">Strong</Text>
      <Text as="u">Misspeled</Text>
    </div>
  ),
};
