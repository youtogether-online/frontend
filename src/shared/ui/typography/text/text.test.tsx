import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import { Text } from "./text";

describe("Text", () => {
  it("Text respects order prop", () => {
    const container = render(<Text>Default text</Text>);
    const text = container.getByText("Default text");
    expect(text).toMatchSnapshot();
  });
});
