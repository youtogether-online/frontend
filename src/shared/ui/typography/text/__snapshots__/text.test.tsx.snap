// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Text > Text respects order prop 1`] = `
<p>
  Default text
</p>
`;
